includes/controllers/class-llms-controller-checkout.php

```diff
@@ -0,0 +1,404 @@
+
+	}
+
+	/**
+	 * AJAX checkout new order controller.
+	 *
+	 * Handles AJAX form submission of the checkout form for new (or pending) orders.
+	 *
+	 * Verifies the AJAX request, passes the `$_POST` data to {@see LLMS_Order_Generator::generate},
+	 * hands the resulting order and data to the gateway's `handle_pending_order()` method and then
+	 * outputs a JSON response object.
+	 *
+	 * @since [version]
+	 *
+	 * @return void
+	 */
+	public function create_pending_order_ajax() {
+
+		$verify = $this->verify_request( self::AJAX_QS_VAR, self::ACTION_CREATE_PENDING_ORDER );
+		if ( ! $verify ) {
+			return $verify;
+		}
+
+		// Ensure the non-ajax handler doesn't also trigger.
+		remove_action( 'init', array( $this, 'create_pending_order' ) );
+
+		// Generate the order.
+		$generator = new LLMS_Order_Generator( $_POST ); // phpcs:ignore WordPress.Security.NonceVerification.Missing -- Nonce verified via `verify_request()`.
+		$order     = $generator->generate( $generator::UA_VALIDATE );
+		if ( is_wp_error( $order ) ) {
+			$this->send_json( $order );
+		}
+
+		// Pending order creation success, pass it over to the gateway.
+		$handle = $generator->get_gateway()->handle_pending_order(
+			$order,
+			$generator->get_plan(),
+			$generator->get_user_data(),
+			$generator->get_coupon()
+		);
+
+		// Automatically add the order key to non-error return arrays.
+		if ( ! is_wp_error( $handle ) ) {
+			$handle['order_key'] = $order->get( 'order_key' );
+		}
+
+		$this->send_json( $handle );
+
+	}
+
+	/**
+	 * Extracts data from `$_POST` into an array that can be passed into `llms_setup_pending_order()`.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $posted_data Data array, from `$_POST`.
+	 * @return array
+	 */
+	private function extract_setup_data( $posted_data ) {
+
+		$plan_id = absint( $posted_data['llms_plan_id'] ?? 0 );
+
+		$data = array(
+			'plan_id'         => $plan_id,
+			'agree_to_terms'  => llms_parse_bool( $posted_data['llms_agree_to_terms'] ?? '' ),
+			'coupon_code'     => sanitize_text_field( $posted_data['llms_coupon_code'] ?? '' ),
+			'customer'        => $this->extract_user_data( $posted_data, $plan_id ),
+			'payment_gateway' => sanitize_text_field( $posted_data['llms_payment_gateway'] ?? '' ),
+		);
+
+		return $data;
+
+	}
+
+	/**
+	 * Extracts user registration / update information from a posted data array.
+	 *
+	 * @since [version]
+	 *
+	 * @param array $posted_data Raw $_POST (or similar) data.
+	 * @return array
+	 */
+	private function extract_user_data( $posted_data, $plan_id ) {
+
+		$user_data = array();
+		$plan      = $plan_id ? llms_get_post( $plan_id ) : false;
+
+		$uid = get_current_user_id();
+		if ( $uid ) {
+			$user_data['user_id'] = $uid;
+		}
+
+		foreach ( LLMS_Forms::instance()->get_form_fields( 'checkout', compact( 'plan' ) ) as $field ) {
+			if ( isset( $posted_data[ $field['name'] ] ) ) {
+				$user_data[ $field['name'] ] = $posted_data[ $field['name'] ];
+			}
+		}
+
+		return $user_data;
```

Comment:
```suggestion
		$plan      = $plan_id ? llms_get_post( $plan_id ) : false;
		$user_data = array_intersect_key( 
			$posted_data, 
			array_flip( array_column( LLMS_Forms::instance()->get_form_fields( 'checkout', compact( 'plan' ) ), 'name' ) )
		);

		$uid = get_current_user_id();
		if ( $uid ) {
			$user_data['user_id'] = $uid;
		}

		return $user_data;
```
what do you think about this golf?